### PR TITLE
Make version segments accessible

### PR DESCRIPTION
**Pull request by mezzodrinker**
This PR is supposed to make all segments of a `ComparableVersion` accessible from the outside while protecting `ComparableVersion.Item` from being extended by clients.

### CODE REVIEW

**Comment by williewillus** [comment]
Why the change to ListItem to compose instead of inherit?
Why this PR in general?
You say you change x, y, z, but it'd be nice to have concrete **use cases** as for why you need this and why other people will benefit from this

**Comment by mezzodrinker** [comment]
@williewillus 

> Why the change to ListItem to compose instead of inherit?

I made ListItem extend Item and implement List because, well, you can't have two superclasses in Java. Also, because if IntegerItem, StringItem and ListItem were to be exposed, that would mean it'd be useful to expose their superclass (or -interface) as well. Because the superclass can't be made `final` for obvious reasons, Item needs to be protected from being extended by clients in a different way. The only way to do that which I know about was to turn Item into an abstract class and then add a checker method which prohibits "custom" subclasses.

> Why this PR in general?
> You say you change x, y, z, but it'd be nice to have concrete _use cases_ as for why you need this and why other people will benefit from this

While working on MobTalker2, I ran into the problem that FML's `ComparableVersion` is able to compare two different versions but it is not possible to access the segments of that version without using reflection. The reason I needed to do that is because MobTalker2 uses a scripting engine to run Lua-like scripts in Minecraft. Since MobTalker2 defines its own "Script API", whose current version may or may not be compatible with the version used in a script, I planned to change the prerequisite of a script having to match the exact Script API version so that scripts which access a version with the same major and lower or equal minor version can still be executed (see [this commit on BitBucket](https://bitbucket.org/MobTalker2/mobtalker2/commits/162512d195d942813e747b13a2602697c34b3a0a)). Now, I know that some other Forge mods also provide scriptable interfaces and may or may not define their own API. I hope you get my idea.

File: src/main/java/net/minecraftforge/fml/common/versioning/ComparableVersion.java [hunk 1]
```diff
@@ -136,16 +161,16 @@ public int compareTo( Item item )
...
+            case STRING_ITEM:
+                return 1; // 1.1 > 1-sp
 
-                case LIST_ITEM:
-                    return 1; // 1.1 > 1-1
+            case LIST_ITEM:
+                return 1; // 1.1 > 1-1
 
-                default:
+            default:
```
Comment (LexManos) [inline src/main/java/net/minecraftforge/fml/common/versioning/ComparableVersion.java]: REALLY!?

Comment (mezzodrinker) [inline src/main/java/net/minecraftforge/fml/common/versioning/ComparableVersion.java]: ...what. Sorry, if I had noticed that, those changes would be long gone. I'm on it.

**Comment by LexManos** [comment]
Your wall of text is rather confusing, why are you trying to dynamically change a version restriction, Just make a new ComparibleVersion.
You do not need access to the internals of the version.
User a proper versioning system like semver and you'll be fine.

Not to mention this PR is horrid as it looks like someone just found the refactor tools of their IDE and went hog wild.

**Pull request closed without merging**